index.bs

```diff
@@ -990,6 +990,184 @@ while (true) {
 }
 </pre>
 
+## Complete example ##  {#example-complete}
+
+*This section is non-normative.*
+
+This example illustrates use of the closed and ready promises, opening
+of uni-directional and bi-directional streams by either the client or
+the server, and sending and receiving datagrams.
+
+<pre class="example" highlight="js">
+// Adds an entry to the event log on the page, optionally applying a specified
+// CSS class.
+
+let currentTransport, streamNumber, currentTransportDatagramWriter;
+
+// "Connect" button handler.
+async function connect() {
+  const url = document.getElementById('url').value;
+  try {
+    var transport = new WebTransport(url);
+    addToEventLog('Initiating connection...');
+  } catch (e) {
+    addToEventLog('Failed to create connection object. ' + e, 'error');
+    return;
+  }
+
+  try {
+    await transport.ready;
+    addToEventLog('Connection ready.');
+  } catch (e) {
+    addToEventLog('Connection failed. ' + e, 'error');
+    return;
+  }
+
+  transport.closed
+      .then(() => {
+        addToEventLog('Connection closed normally.');
+      })
+      .catch(() => {
+        addToEventLog('Connection closed abruptly.', 'error');
+      });
+
+  currentTransport = transport;
+  streamNumber = 1;
+  try {
+    currentTransportDatagramWriter = transport.datagramWritable.getWriter();
+    addToEventLog('Datagram writer ready.');
+  } catch (e) {
+    addToEventLog('Sending datagrams not supported: ' + e, 'error');
+    return;
+  }
+  readDatagrams(transport);
+  acceptUnidirectionalStreams(transport);
+  document.forms.sending.elements.send.disabled = false;
+  document.getElementById('connect').disabled = true;
+}
+
+// "Send data" button handler.
+async function sendData() {
+  let form = document.forms.sending.elements;
+  let encoder = new TextEncoder('utf-8');
+  let rawData = sending.data.value;
+  let data = encoder.encode(rawData);
+  let transport = currentTransport;
+  try {
+    switch (form.sendtype.value) {
+      case 'datagram':
+        await currentTransportDatagramWriter.write(data);
+        addToEventLog('Sent datagram: ' + rawData);
+        break;
+      case 'unidi': {
+        let stream = await transport.createUnidirectionalStream();
+        let writer = stream.writable.getWriter();
+        await writer.write(data);
+        writer.close();
+        addToEventLog('Sent a unidirectional stream with data: ' + rawData);
+        break;
+      }
+      case 'bidi': {
+        let stream = await transport.createBidirectionalStream();
+        let number = streamNumber++;
+        readFromIncomingStream(stream, number);
+
+        let writer = stream.writable.getWriter();
+        await writer.write(data);
+        writer.close();
+        addToEventLog(
+            'Opened bidirectional stream #' + number +
```

Comment:
This is likely problematic: the stream number here is not the stream ID from QUIC (or h2).  That has the potential to create some really nasty confusion.